pal: Add support for displayable fp16 swapchain format.

kleinerm opened the pull request:
Maps VK_FORMAT_R16_G16_B16_A16_SFLOAT to amdgpu-kms format
DRM_FORMAT_XBGR16161616F. Linux 5.8 amdgpu-kms supports
this format for scanout and display on gpu's with DCE-11.2
display engine (Polaris family) and later generations.

Linux kernel 5.8 mainline commit for this new feature is
2a5195dca0b7058e65443416a20680be985d8753
("drm/amd/display: Expose support for xBGR ordered fp16 formats.")

Successfully tested on Polaris DCE-11.2 and Raven Ridge DCN-1.

Signed-off-by: Mario Kleiner <mario.kleiner.de@gmail.com>

amdvlk-admin commented:
Can one of the admins verify this patch?

JacobHeAMD commented:
ok to test

JacobHeAMD commented:
test this please

JaxLinAMD commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 980289e4a1d5eac320fe19fd664a5900e45c268e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>LLPC: dev d0d1016eee8ef5d32f73fa5da58c30c8a3e1aea7</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev f4a992dd7e556ed5e7e2ffa2c830f1cd79bd4596</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c2ea9a11546d054e3918c44ac3b3ad930686cd79</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/130900)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22784/43839 (52.0%)</li>
<li>Failed: 0/43839 (0.0%)</li>
<li>Skipped: 21055/43839 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22783/43843 (52.0%)</li>
<li>Failed: 0/43843 (0.0%)</li>
<li>Skipped: 21060/43843 (48.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21609/43218 (50.0%)</li>
<li>Failed: 0/43218 (0.0%)</li>
<li>Skipped: 21609/43218 (50.0%)</li>
</ul></details>


</ul></details>

JacobHeAMD commented on src/core/os/amdgpu/amdgpuDevice.cpp:
```diff
@@ -1986,6 +2005,24 @@ Result Device::CreateInternalImage(
+    // and later.
+    if ((IsDrmVersionOrGreater(3, 38) || IsKernelVersionEqualOrGreater(5, 8)) &&
+        (IsGfx10Plus(m_chipProperties.gfxLevel) || IsGfx9(*this) ||
+        (IsGfx8(*this) && (IsPolaris10(*this) || IsPolaris11(*this) || IsPolaris12(*this))))) {
```
"{" should occupy one new line with proper alignment.

kleinerm replied:
Thanks. Done, in just updated pull request.

JacobHeAMD commented on src/core/os/amdgpu/amdgpuDevice.cpp:
```diff
@@ -2050,11 +2089,25 @@ Result Device::GetSwapChainInfo(
+        pSwapChainProperties->imageFormatCount = baseFormatCount;
+
+        // Some gpu + amdgpu kms combinations do support fp16 scanout and display.
+        if (HasFP16DisplaySupport()) {
```
ditto

kleinerm replied:
Thanks. Done, in just updated pull request.

kleinerm commented:
I  changed the order of commits in this force-pushed update, as the  'max bpc' patch seems to be fine as is. Other than the code style fix, the patches are the same. Thanks.

JacobHeAMD commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 66a81660022bbbb207ec63c5ab6f0c5b5703f121
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>LLPC: dev 26120c68e0195722784e496b05e6ab57e21642c1</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev f4a992dd7e556ed5e7e2ffa2c830f1cd79bd4596</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ea979e90c306a6ddec5db687e9c8d33b5289a23c</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131499)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22784/43839 (52.0%)</li>
<li>Failed: 0/43839 (0.0%)</li>
<li>Skipped: 21055/43839 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22783/43843 (52.0%)</li>
<li>Failed: 0/43843 (0.0%)</li>
<li>Skipped: 21060/43843 (48.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21609/43817 (49.3%)</li>
<li>Failed: 0/43817 (0.0%)</li>
<li>Skipped: 22208/43817 (50.7%)</li>
</ul></details>


</ul></details>

RobMartin-AMD commented on src/core/os/amdgpu/amdgpuDevice.cpp:
```diff
@@ -1986,6 +2005,25 @@ Result Device::CreateInternalImage(
 
+// =====================================================================================================================
+// Returns true if gpu + amdgpu kms driver do support 16 bit floating point display.
+bool Device::HasFP16DisplaySupport(void)
```
Coding guidelines:  
"FP16" should instead be "Fp16".  I know its odd, but we handle the UpperCamelCase for acronyms and initialisms the same as we do for normal words.

Also, the "void" in the argument list should be omitted.  Our style guidelines don't include that.

kleinerm replied:
Done in force-push just now.

For efficiency, it would be helpful to external contributors if these style guidelines would be publicly available, unless i missed them?

RobMartin-AMD commented on src/core/os/amdgpu/amdgpuDevice.cpp:
```diff
@@ -2050,11 +2090,26 @@ Result Device::GetSwapChainInfo(
+            // fp16 is first slot in Presentable16BitSwizzledFormat[].
+            pSwapChainProperties->imageFormatCount++;
+        }
+
```
Please add a PAL_ASSERT that pSwapChainProperties->imageFormatCount does not exceed the MaxPresentableImageFormat constant defined in palSwapChain.h.

kleinerm replied:
Done in force-push just now.

kleinerm commented:
Style fixes and adding an assert done in updated pull request. Thanks for the review.

kleinerm commented:
@RobMartin-AMD ping.

JaxLinAMD commented:
retest this please

kleinerm commented:
@RobMartin-AMD @JacobHeAMD ping? Would be nice to get this squared away, given that nobody voiced any objections against the general approach and implementation, only coding style nits. Thanks.

JacobHeAMD commented:
@JaxLinAMD , please help to merge this pull request.

JaxLinAMD commented:
retest this please

amdvlk-admin commented:
### Test summary for commit e68b4ef46b8de7cfda99d0c37ff0f2273852e86b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>LLPC: dev f23bb234ee3ad79aad078995012edf93f8938676</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev f4a992dd7e556ed5e7e2ffa2c830f1cd79bd4596</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ccbc834ac4f92e63a9a003bd5a18cb9fcf3c6121</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131507)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22784/43840 (52.0%)</li>
<li>Failed: 0/43840 (0.0%)</li>
<li>Skipped: 21056/43840 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22783/43844 (52.0%)</li>
<li>Failed: 0/43844 (0.0%)</li>
<li>Skipped: 21061/43844 (48.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21609/43823 (49.3%)</li>
<li>Failed: 0/43823 (0.0%)</li>
<li>Skipped: 22214/43823 (50.7%)</li>
</ul></details>


</ul></details>

JaxLinAMD commented:
@kleinerm , could you add ModeSet back? this should not be removed

JaxLinAMD commented:
@kleinerm also, we prefer rebase than merge. So could you rebase your personal branch on top of dev branch, recover "ModeSet" I deleted and force-push to your branch again? you can also squash those commits into one single commit, that would make git history easy to understand. thanks!

kleinerm commented:
I can try to do that, after some sleep. But do you mean squashing all of my commits into one? Even though the one that fixed 'max bpc' and the other one for fp16 are logically separate things - related, but independent?

JaxLinAMD commented:
Uhm, if they are independent two commits, then no need to squash:)

kleinerm commented:
Ok. I guess i will keep the two patches separate, but add the ModeSet call back to the fp16 patch, so pulling this will also fix that bug from issue #46 i guess that's fine as you are assigned to that issue anyway, so we won't get new merge conflicts if i add it back here.

JaxLinAMD commented:
@kleinerm  FYI, I just submmited https://github.com/GPUOpen-Drivers/pal/commit/bebb2b0d0d70625d825cd386a1360154709487f2
add back "ModeSet(pImage)" you can go on with rebase

JaxLinAMD commented:
@kleinerm  Oops, please rebase and force-push again

kleinerm commented:
@JaxLinAMD Ok, rebased, force-pushed, retested. Ready to pull. I go to bed now.

amdvlk-admin commented:
### Test summary for commit 4b36745ccea3cd44cef9c47192751dc5b0b266e9
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>LLPC: dev f18670dec0c800dd9a7ed514ced8843c4694217a</li>
<li>SPVGEN: dev 4865b574184e261d87e77277282e90e3be14c53a</li>
<li>XGL: dev 598c6832a4983f5b75b38a589fca5be80a2f3bb0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ccbc834ac4f92e63a9a003bd5a18cb9fcf3c6121</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131444)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22766/43819 (52.0%)</li>
<li>Failed: 0/43819 (0.0%)</li>
<li>Skipped: 21053/43819 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22765/43823 (51.9%)</li>
<li>Failed: 0/43823 (0.0%)</li>
<li>Skipped: 21058/43823 (48.1%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21591/43802 (49.3%)</li>
<li>Failed: 0/43802 (0.0%)</li>
<li>Skipped: 22211/43802 (50.7%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit a4690886ce42835c3da2c0fe8af9690579776aff
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>LLPC: dev f18670dec0c800dd9a7ed514ced8843c4694217a</li>
<li>SPVGEN: dev 4865b574184e261d87e77277282e90e3be14c53a</li>
<li>XGL: dev 598c6832a4983f5b75b38a589fca5be80a2f3bb0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ccbc834ac4f92e63a9a003bd5a18cb9fcf3c6121</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131444)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22766/43819 (52.0%)</li>
<li>Failed: 0/43819 (0.0%)</li>
<li>Skipped: 21053/43819 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22765/43823 (51.9%)</li>
<li>Failed: 0/43823 (0.0%)</li>
<li>Skipped: 21058/43823 (48.1%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21591/43802 (49.3%)</li>
<li>Failed: 0/43802 (0.0%)</li>
<li>Skipped: 22211/43802 (50.7%)</li>
</ul></details>


</ul></details>

JaxLinAMD commented:
@kleinerm , Have a good night! Shouldn't it be " ret = ModeSet(pImage) != Result::Success;"? I guess you may want to change it again, currently it is" ModeSet(pImage)"
'

kleinerm commented:
@JaxLinAMD No, ModeSet(pImage) is fine for now, doesn't really matter for this series, was just some small refinement. I don't want to go through another modify/compile/test cycle and delay this pull any longer.

kleinerm commented:
@JaxLinAMD @JacobHeAMD Ok, got some sleep, changed my mind, added "ret = ModeSet(pImage) != Result::Success;" back into the commit. It doesn't really matter for the series, but a little improvement in error handling never hurts.

Retested, rebased, ready to pull. Please please can we get this in now before some random other commits from your side break it again and produce new merge conflicts?

Thanks.

JaxLinAMD commented:
retest this please

kleinerm commented:
Thanks!
